content/docs/fuzzing/c-cpp/techniques/01-coverage/index.md

```diff
@@ -0,0 +1,469 @@
+
+Alternatively, filter out crashing inputs manually or automatically. Another way might be to `fork()` the process before executing the SUT (i.e., before calling `LLVMFuzzerTestOneInput`).
+
+If you want to generate an HTML report using the LLVM stack, then we recommend exporting to the lcov format and then using the lcov tool to generate an HTML page. Using lcov, you can generate a per-source-file report, which is not possible with LLVM before version 18.
+
+```shell
+llvm-cov export ./fuzz_exec -instr-profile=fuzz.profdata -ignore-filename-regex='harness.cc|execute-rt.cc' -format=lcov > fuzz.lcov
+
+apt install lcov
+genhtml --output-directory fuzz_html/ fuzz.lcov
+```
+
+The resulting HTML is stored in the `fuzz_html/` directory.
+
+{{< resourceFigure "cov2.png" >}}
+HTML coverage report generated by lcov
+{{< /resourceFigure >}}
+
+
+## Code coverage using gcov and gcovr {#code-coverage-using-gcov-and-gcovr}
+
+The [gcov](https://gcc.gnu.org/onlinedocs/gcc/Gcov.html) instrumentation is provided by both GCC and LLVM. Using gcov, we can use the excellent [gcovr](https://gcovr.com/) tool to generate HTML reports. If using LLVM, the GCC instrumentation can be added using the flags `-ftest-coverage` and `-fprofile-arcs`. Using the [`execute-rt.cc`](#execute_rt_src) runtime from the previous section, we can compile the coverage binary.
+
+
+```shell
+clang++ -DNO_MAIN -O2 -ftest-coverage -fprofile-arcs main.cc harness.cc execute-rt.cc -o fuzz_exec_gcov
+```
+
+
+The same compilation instruction is also valid for GCC. In this case, replace `clang++` with `g++`.
+
+During compilation, `.gnco` files are created for each object file. During execution, `.gnca` files are created or updated for each object file. The environment variables `GCOV_PREFIX` and `GCOV_PREFIX_STRIP` can be used to copy files across (see [here](https://llvm.org/docs/CommandGuide/llvm-cov.html#id2:~:text=By%20default%2C%20the,to%20find%20them.) for more information). For every invocation of the binary, the .gnca files are incrementally updated. This means that executing the program over and over without removing old .gnca files will continuously increment hit counters. The gcovr tool simplifies the handling of these files.
+
+The first step is to execute the coverage binary on an existing corpus.
+
+
+```shell
+./fuzz_exec_gcov corpus/
+```
+
+We now make sure that gcovr is installed. We recommend installing gcovr through Python's pip because distribution versions could be outdated. If you use LLVM/Clang 14 or above, at least gcovr 5.1 is required (the verison 5.1 contains this [patch](https://github.com/gcovr/gcovr/pull/525/files#diff-26284e367af8a1344772d5e18ab13d7e5518c701936c21733b2a18c13e2b958cL31-R31)).
```

Comment:
```suggestion
We now make sure that gcovr is installed. We recommend installing gcovr through Python's pip because distribution versions could be outdated. If you use LLVM/Clang 14 or above, at least gcovr 5.1 is required (the version 5.1 contains this [patch](https://github.com/gcovr/gcovr/pull/525/files#diff-26284e367af8a1344772d5e18ab13d7e5518c701936c21733b2a18c13e2b958cL31-R31)).
```